File: backend/app/schemas/shubhali_holat.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.foydalanuvchi import Smena
from app.models.shubhali_holat import ShubhaliHolatStatusi


class ShubhaliHolatJavob(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    vaqt: datetime
    smena: Smena | None
    ogirlik: float
    surat_yoli: str | None
    holati: ShubhaliHolatStatusi
    korib_chiqqan_id: int | None
    korib_chiqilgan_vaqt: datetime | None
    stansiya_id: int | None = None


class ShubhaliHolatRoyxatJavob(ShubhaliHolatJavob):
    korib_chiqqan_ism: str | None = None
    operator_ism: str | None = None
    tasdiqlangan: bool


class ShubhaliHolatKipSifatidaSaqlash(BaseModel):
    """`POST /shubhali-holatlar/{id}/saqlash` so'rov tanasi — admin hodisani
    HAQIQIY Kip sifatida saqlash uchun mahsulot/partiyani qo'lda tanlaydi
    (kip-to'g'irlash zayavkasi yaratishdagi bilan bir xil naqsh)."""

    mahsulot_kodi: str
    partiya_raqami: int


class ShubhaliOperatorSoni(BaseModel):
    operator_id: int
    ism: str
    soni: int


class ShubhaliHolatStatistika(BaseModel):
    smena_boyicha: dict[str, int]
    operator_boyicha: list[ShubhaliOperatorSoni]

File: frontend/lib/models/shubhali_holat.dart
class ShubhaliHolat {
  final int id;
  final DateTime vaqt;
  final String? smena;
  final double ogirlik;
  final String? suratYoli;
  final String holati;
  final bool tasdiqlangan;
  final int? koribChiqqanId;
  final DateTime? koribChiqilganVaqt;
  final String? koribChiqqanIsm;
  final String? operatorIsm;

  ShubhaliHolat({
    required this.id,
    required this.vaqt,
    this.smena,
    required this.ogirlik,
    this.suratYoli,
    required this.holati,
    required this.tasdiqlangan,
    this.koribChiqqanId,
    this.koribChiqilganVaqt,
    this.koribChiqqanIsm,
    this.operatorIsm,
  });

  factory ShubhaliHolat.fromJson(Map<String, dynamic> j) => ShubhaliHolat(
        id: j['id'],
        vaqt: DateTime.parse(j['vaqt']),
        smena: j['smena'],
        ogirlik: (j['ogirlik'] as num).toDouble(),
        suratYoli: j['surat_yoli'],
        holati: j['holati'],
        tasdiqlangan: j['tasdiqlangan'] ?? false,
        koribChiqqanId: j['korib_chiqqan_id'],
        koribChiqilganVaqt: j['korib_chiqilgan_vaqt'] == null ? null : DateTime.parse(j['korib_chiqilgan_vaqt']),
        koribChiqqanIsm: j['korib_chiqqan_ism'],
        operatorIsm: j['operator_ism'],
      );
}

class ShubhaliHolatStatistika {
  final Map<String, int> smenaBoyicha;

  ShubhaliHolatStatistika({required this.smenaBoyicha});

  factory ShubhaliHolatStatistika.fromJson(Map<String, dynamic> j) => ShubhaliHolatStatistika(
        smenaBoyicha: (j['smena_boyicha'] as Map<String, dynamic>).map((k, v) => MapEntry(k, v as int)),
      );
}
